helix-core/src/main/java/org/apache/helix/model/InstanceConfig.java

```diff
@@ -264,9 +264,22 @@ public boolean getInstanceEnabled() {
    * @param enabled true to enable, false to disable
    */
   public void setInstanceEnabled(boolean enabled) {
+    // set instance operation only when we need to change InstanceEnabled value.
+    // When enabling an instance where HELIX_ENABLED is false, we update INSTANCE_OPERATION to 'ENABLE'
+    // When disabling and instance where HELIX_ENABLED is false, we overwrite what current operation and
+    // update INSTANCE_OPERATION to 'DISABLE'.
+    String instanceOperationKey = InstanceConfigProperty.INSTANCE_OPERATION.toString();
+    if (enabled != getInstanceEnabled() && _record.getSimpleField(instanceOperationKey) != null) {
```

Comment:
!getInstanceEnabled()

Reply:
Had an offline discussion. We update both operation and helix-enabled when current enabled is not the same as provided value. We also update operation regardless of if there is existing value or not.